include/drivers/pcie/msi.h

```diff
@@ -7,70 +7,138 @@
 #ifndef ZEPHYR_INCLUDE_DRIVERS_PCIE_MSI_H_
 #define ZEPHYR_INCLUDE_DRIVERS_PCIE_MSI_H_
 
-#include <drivers/pcie/pcie.h>
+#include <kernel.h>
 #include <zephyr/types.h>
 #include <stdbool.h>
 
+#include <drivers/pcie/pcie.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+struct msix_vector {
+	uint32_t msg_addr;
+	uint32_t msg_up_addr;
+	uint32_t msg_data;
+	uint32_t vector_ctrl;
+};
+
+struct msi_vector {
+	pcie_bdf_t bdf;
+	arch_msi_vector_t arch;
+#ifdef CONFIG_PCIE_MSI_X
+	struct msix_vector *msix_vector;
+	bool msix;
+#endif /* CONFIG_PCIE_MSI_X */
+};
+
+typedef struct msi_vector msi_vector_t;
+
+#ifdef CONFIG_PCIE_MSI_MULTI_VECTOR
+
 /**
- * @brief Find a PCI(e) capability in an endpoint's configuration space.
+ * @brief Allocate vector(s) for the endpoint MSI message(s)
+ * This is arch-specific.
  *
- * @param bdf the PCI endpoint to examine
- * @param cap_id the capability ID of interest
- * @return the index of the configuration word, or 0 if no capability.
+ * @param priority the MSI vectors base interrupt priority
+ * @param vectors an array to fill with allocated MSI vectors
+ * @param n_vector the size of MSI vectors array
  *
- * Note: PCI(e) capabilities are only used in the MSI code, so for
- * now, capabilities-related code is only included when MSI is. It
- * can easily be separated out if/when its use spreads.
+ * @return The number of allocated MSI vectors
  */
-extern uint32_t pcie_get_cap(pcie_bdf_t bdf, uint32_t cap_id);
+uint8_t arch_allocate_msi_vectors(unsigned int priority,
```

Comment:
do not define `arch_` interfaces anywhere other than include/sys/arch_interface.h or kernel/include/kernel_arch_interface.h